<a href="https://colab.research.google.com/github/Kunyarat4803/Python_dataviz2024/blob/main/Data_Viz_2024_Exam2_653020567_7_%E0%B8%81%E0%B8%B1%E0%B8%A5%E0%B8%A2%E0%B8%A3%E0%B8%B1%E0%B8%95%E0%B8%99%E0%B9%8C_%E0%B8%AD%E0%B8%B4%E0%B8%99%E0%B8%97%E0%B8%A3%E0%B9%8C%E0%B8%AA%E0%B8%B3%E0%B8%A3%E0%B8%B2%E0%B8%8D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ให้สืบหาฆาตรกร ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง SQL City โดยเริ่มจากการตรวจสอบ ข้อมูลคดี (crime_scene_report.csv)

หมายเหตุ
* person.csv ข้อมูลประชากร
* interview.csv ข้อมูลการสอบสวน
* income.csv ข้อมูลรายได้
* get_fit_now_member.csv ข้อมูลสมาชิก ฟิตเนส Get Fit Now
* get_fit_now_check_in.csv ข้อมูลการเข้าใช้ ฟิตเนส Get Fit Now
* facebook_event_checkin.csv ข้อมูลการเช็คอิน Facebook
* driver_license.csv ข้อมูลใบขับขี่
* crime_scene_report.csv ข้อมูลคดี



In [1]:
#นำเข้าข้อมูล
import pandas
import os
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import os

path = '/content/drive/MyDrive/data_viz_2024_Data/Exam_2'
person_path = os.path.join(path,'person.csv')
print(person_path)
interview_path = os.path.join(path,'interview.csv')
print(interview_path)
Income_path = os.path.join(path,'income.csv')
print(Income_path)
get_fit_now_member_path = os.path.join(path,'get_fit_now_member.csv')
print(get_fit_now_member_path)
get_fit_now_check_in_path = os.path.join(path,'get_fit_now_check_in.csv')
print(get_fit_now_check_in_path)
facebook_event_checkin_path = os.path.join(path,'facebook_event_checkin.csv')
print(facebook_event_checkin_path)
drivers_license_path = os.path.join(path,'drivers_license.csv')
print(drivers_license_path)
crime_scene_report_path = os.path.join(path,'crime_scene_report.csv')
print(crime_scene_report_path)

/content/drive/MyDrive/data_viz_2024_Data/Exam_2/person.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/interview.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/income.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/get_fit_now_member.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/get_fit_now_check_in.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/facebook_event_checkin.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/drivers_license.csv
/content/drive/MyDrive/data_viz_2024_Data/Exam_2/crime_scene_report.csv


In [5]:
person = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/person.csv')
person



id                 name  license_id  address_number  \
0      10000   Christoper Peteuil      993845             624   
1      10007  Kourtney Calderwood      861794            2791   
2      10010            Muoi Cary      385336             741   
3      10016          Era Moselle      431897            1987   
4      10025         Trena Hornby      550890             276   
...      ...                  ...         ...             ...   
10006  99936          Luba Benser      274427             680   
10007  99941      Roxana Mckimley      975942            1613   
10008  99965      Cherie Zeimantz      287627            3661   
10009  99982          Allen Cruse      251350            3126   
10010  99990         Vance Hunten      830407            3056   

      address_street_name        ssn  
0            Bankhall Ave  747714076  
1           Gustavus Blvd  477972044  
2         Northwestern Dr  828638512  
3           Wood Glade St  614621061  
4           Daws Hill Way  223877684  
...                   ...        ...  
10006        Carnage Blvd  685095054  
10007             Gate St  512136801  
10008       The Water Ave  362877324  
10009           N Jean Dr  348734531  
10010       Lancefield St  896677562  

[10011 rows x 6 columns]

In [6]:
interview = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/interview.csv')
interview

person_id                                         transcript
0         28508                ‘I deny it!’ said the March Hare.\n
1         63713                                                 \n
2         86208      way, and the whole party swam to the shore.\n
3         35267  lessons in here? Why, there’s hardly room for ...
4         33856                                                 \n
...         ...                                                ...
4986      37357  Alice did not wish to offend the Dormouse agai...
4987      10206  time,’ she said, ‘than waste it in asking ridd...
4988      14887  I heard a gunshot and then saw a man run out. ...
4989      16371  I saw the murder happen, and I recognized the ...
4990      67318  I was hired by a woman with a lot of money. I ...

[4991 rows x 2 columns]

In [7]:
income = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/income.csv')
income

ssn  annual_income
0     100009868          52200
1     100169584          64500
2     100300433          74400
3     100355733          35900
4     100366269          73000
...         ...            ...
7509  999679296          54400
7510  999762859          77000
7511  999824984          82000
7512  999910617          82600
7513  999942603          11500

[7514 rows x 2 columns]

In [8]:
get_fit_now_member = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/get_fit_now_member.csv')
get_fit_now_member

id  person_id                name  membership_start_date  \
0    NL318      65076     Everette Koepke               20170926   
1    AOE21      39426        Noe Locascio               20171005   
2    2PN28      63823  Jeromy Heitschmidt               20180215   
3    0YJ24      80651      Waneta Wellard               20171206   
4    3A08L      32858        Mei Bianchin               20170401   
..     ...        ...                 ...                    ...   
179  2V137      41693      Wendell Dulany               20171219   
180  4KB72      79110          Emile Hege               20170522   
181  48Z7A      28819        Joe Germuska               20160305   
182  48Z55      67318       Jeremy Bowers               20160101   
183  90081      16371      Annabel Miller               20160208   

    membership_status  
0                gold  
1             regular  
2              silver  
3                gold  
4              silver  
..                ...  
179            silver  
180           regular  
181              gold  
182              gold  
183              gold  

[184 rows x 5 columns]

In [9]:
get_fit_now_check_in = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/get_fit_now_check_in.csv')
get_fit_now_check_in

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [10]:
facebook_event_checkin = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/facebook_event_checkin.csv')
facebook_event_checkin

person_id  event_id                                         event_name  \
0          28508      5880    Nudists are people who wear one-button suits.\n   
1          63713      3865  but that's because it's the best book on anyth...   
2          63713      3999           If Murphy's Law can go wrong, it will.\n   
3          63713      6436  Old programmers never die.  They just branch t...   
4          82998      4470               Help a swallow land at Capistrano.\n   
...          ...       ...                                                ...   
20006      99716      1143                               SQL Symphony Concert   
20007      99716      1143                               SQL Symphony Concert   
20008      99716      1143                               SQL Symphony Concert   
20009      67318      4719                             The Funky Grooves Tour   
20010      67318      1143                               SQL Symphony Concert   

           date  
0      20170913  
1      20171009  
2      20170502  
3      20170926  
4      20171022  
...         ...  
20006  20171206  
20007  20171212  
20008  20171229  
20009  20180115  
20010  20171206  

[20011 rows x 4 columns]

In [11]:
driver_license = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/drivers_license.csv')
driver_license

id  age  height eye_color hair_color  gender plate_number  \
0      100280   72      57     brown        red    male       P24L4U   
1      100460   63      72     brown      brown  female       XF02T6   
2      101029   62      74     green      green  female       VKY5KR   
3      101198   43      54     amber      brown  female       Y5NZ08   
4      101255   18      79      blue       grey  female       5162Z1   
...       ...  ...     ...       ...        ...     ...          ...   
10002  999923   19      77     amber      black  female       5L0ZI4   
10003  999940   71      61     green      green    male       1B8QN8   
10004  999981   67      69     brown       blue  female       1684K3   
10005  999986   49      58     green       grey    male       F8F64H   
10006  999993   18      63     black      black  female       6UZO2O   

         car_make    car_model  
0           Acura          MDX  
1        Cadillac          SRX  
2           Scion           xB  
3          Nissan        Rogue  
4           Lexus           GS  
...           ...          ...  
10002         GMC  Sierra 3500  
10003  Mitsubishi      Eclipse  
10004  Land Rover          LR2  
10005       Lexus           LS  
10006    Cadillac      DeVille  

[10007 rows x 9 columns]

เริ่มการสืบส่วน

In [12]:
# prompt: หาคดีฆาตกรรม (murder) ในวันที่ 15 มกราคม 2018 ที่เมือง SQL City
import pandas as pd
# ดึงข้อมูล crime_scene_report.csv
crime_scene_report_df = pd.read_csv('/content/drive/MyDrive/data_viz_2024_Data/Exam_2/crime_scene_report.csv')
# คัดเอาเฉพาะข้อมูลที่เราต้องการ
# คือ คดีฆาตกรรม (murder) ในวันที่ 15 มกราคม 2018 (20180115) ที่เมือง SQL City
select_data = crime_scene_report_df[(crime_scene_report_df['date'] == 20180115) & (crime_scene_report_df['type'] == 'murder') & (crime_scene_report_df['city'] == 'SQL City')]

# แสดงผลลัพธ์
select_data

date    type                                        description  \
1227  20180115  murder  Security footage shows that there were 2 witne...   

          city  
1227  SQL City

วิเคราะห์โจทย์ที่ได้จากการหาคดีฆ่าตกรรม ในวันที่ 15 มกราคม  2018 ที่เมือง SQL City

'Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave.
กล้องวงจรปิดแสดงว่ามีพยาน 2 คน คนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr" ส่วนพยานคนที่สองชื่อ Annabel อาศัยอยู่ที่ใดที่หนึ่งบนถนน "Franklin Ave

#หาข้อมูลพยานคนแรกที่อยู่บ้านหลังสุดท้ายบนถนน Northwestern Dr."

In [13]:
# prompt: ตรวจสอบที่อยู่บ้านบนถนน Northwestern Dr. และค้นหาบ้านที่มีหมายเลขบ้านสูงสุด ซึ่งน่าจะเป็นพยานคนแรก

# ตรวจสอบที่อยู่บ้านบนถนน Northwestern Dr.
northwestern_dr_residents = person[person['address_street_name'] == 'Northwestern Dr']

# ค้นหาบ้านที่มีหมายเลขบ้านสูงสุด
witness1_person_id = northwestern_dr_residents['address_number'].idxmax()
witness1 = person.loc[witness1_person_id]

# แสดงข้อมูลพยานคนแรก
witness1

id                               14887
name                    Morty Schapiro
license_id                      118009
address_number                    4919
address_street_name    Northwestern Dr
ssn                          111564949
Name: 499, dtype: object

เมื่อทราบว่าพยานคนที่ 1 คือ Morty Schapiro ที่มี id หรือ person id คือ 14887 เราสามารถนำไปหาข้อมูลการสอบสวนได้

In [14]:
# prompt: เราสามารถนำไปหาข้อมูลการสอบสวนได้

# หาข้อมูลการสอบสวนของพยานคนที่ 1 (Morty Schapiro)
witness1_interview = interview[interview['person_id'] == 14887]

# แสดงข้อมูลการสอบสวน
print(witness1_interview)


      person_id                                         transcript
4988      14887  I heard a gunshot and then saw a man run out. ...


In [15]:
# prompt: ดูข้อความใน column transcript ทั้งหมดของwitness1_interview

# ดูข้อความใน column transcript ทั้งหมดของwitness1_interview
for index, row in witness1_interview.iterrows():
    print(row['transcript'])


I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".


ฉันได้ยินเสียงปืนแล้วก็เห็นชายคนหนึ่งวิ่งออกไป เขามีกระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" เฉพาะสมาชิกระดับโกลด์เท่านั้นที่มีถุงเหล่านั้น ชายคนนั้นขึ้นรถพร้อมจานที่มี "H42W"

#หาข้อมูลเกี่ยวกับพยานคนที่สอง ชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนน "Franklin Ave"

In [16]:
# prompt:  หา person ที่ชื่อ Annabel และอยู่  "Franklin Ave"

# หา person ที่ชื่อ Annabel และอยู่  "Franklin Ave"
witness2 = person[(person['address_street_name'] == 'Franklin Ave') & (person['name'].str.contains('Annabel'))]
witness2


id            name  license_id  address_number address_street_name  \
665  16371  Annabel Miller      490173             103        Franklin Ave   

           ssn  
665  318771143

ทราบว่าพยานคนที่ 2 คือ Annabel Miller ที่มี id หรือ person id คือ 16371 เราสามารถนำไปหาข้อมูลการสอบสวนได้

In [17]:
# prompt:  เราสามารถนำไปหาข้อมูลการสอบสวนได้

# หาข้อมูลการสอบสวนของพยานคนที่ 2 (Annabel Miller)
witness2_interview = interview[interview['person_id'] == 16371]

# แสดงข้อมูลการสอบสวน
print(witness2_interview)

# ดูข้อความใน column transcript ทั้งหมดของwitness2_interview
for index, row in witness2_interview.iterrows():
    print(row['transcript'])


      person_id                                         transcript
4989      16371  I saw the murder happen, and I recognized the ...
I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.


ฉันเห็นการฆาตกรรมเกิดขึ้น และฉันก็จำฆาตกรได้จากยิมของฉันตอนที่ฉันกำลังออกกำลังกายเมื่อสัปดาห์ที่แล้วในวันที่ 9 มกราคม

#เริ่มสืบค้นข้อมูลจากทั้งพยานคนที่ 1 และ 2

In [24]:
#หาว่าใครมีสมาชิกฟิตเนสเป็นระดับ gold และมีรหัสสมาชิกขึ้นต้นด้วย 48Z
gold_member = get_fit_now_member[get_fit_now_member['membership_status'] == 'gold']
gold_member = gold_member[gold_member['id'].str.startswith('48Z')]
gold_member

id  person_id           name  membership_start_date membership_status
181  48Z7A      28819   Joe Germuska               20160305              gold
182  48Z55      67318  Jeremy Bowers               20160101              gold

In [25]:
# prompt: หารายละเอียดของ Joe Germuska และ Jeremy Bowers ในข้อมูลประชากร

# หารายละเอียดของ Joe Germuska และ Jeremy Bowers ในข้อมูลประชากร
suspect_detail = person[(person['name'] == 'Joe Germuska') | (person['name'] == 'Jeremy Bowers')]
suspect_detail


id           name  license_id  address_number    address_street_name  \
2037  28819   Joe Germuska      173289             111                Fisk Rd   
6327  67318  Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

            ssn  
2037  138909730  
6327  871539279

In [26]:
# prompt: หาว่ารถที่มีป้ายทะเบียนรถประกอบด้วย "H42W"

# หารถที่มีป้ายทะเบียนรถประกอบด้วย "H42W"
suspect_car = driver_license[driver_license['plate_number'].str.contains('H42W')]
suspect_car


id  age  height eye_color hair_color  gender plate_number  \
915   183779   21      65      blue     blonde  female       H42W0X   
3529  423327   30      70     brown      brown    male       0H42W2   
6240  664760   21      71     black      black    male       4H42WR   

       car_make car_model  
915      Toyota     Prius  
3529  Chevrolet  Spark LS  
6240     Nissan    Altima

In [27]:
# prompt: หาว่าใครเข้าใช้ฟิตเนตวันที่ 9 มกราคม 2018 และมีรายชื่อตรงกับสมาชิกฟิตเนสเป็นระดับ gold และมีรหัสสมาชิกขึ้นต้นด้วย 48Z

import pandas as pd
# หาว่าใครเข้าใช้ฟิตเนตวันที่ 9 มกราคม 2018
get_fit_now_check_in['check_in_date'] = pd.to_datetime(get_fit_now_check_in['check_in_date'], format='%Y%m%d')
check_in_jan9 = get_fit_now_check_in[get_fit_now_check_in['check_in_date'] == '2018-01-09']

# หาว่าใครมีรายชื่อตรงกับสมาชิกฟิตเนสเป็นระดับ gold และมีรหัสสมาชิกขึ้นต้นด้วย 48Z
suspects = check_in_jan9[check_in_jan9['membership_id'].isin(gold_member1['id'])]

# แสดงรายชื่อผู้ต้องสงสัย
print(suspects)


     membership_id check_in_date  check_in_time  check_out_time
2700         48Z7A    2018-01-09           1600            1730
2701         48Z55    2018-01-09           1530            1700


สรุปจากการวิเคราะห์ข้อมูลและการเชื่อมโยงหลักฐาน
---
ฆาตรกร ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง SQL City คือ **Jeremy Bowers**


*  Jeremy Bowers มี person id 67318 membership 48Z55 (สมาชิกฟิตเนส Get Fit Now)
license_id 423327 อายุ: 30 ปี
ความสูง: 70 นิ้ว
สีตา: น้ำตาล
สีผม: น้ำตาล
เพศ: ชาย
หมายเลขทะเบียนรถ: 0H42W2
ยี่ห้อรถ: Chevrolet
รุ่นรถ: Spark L


*  หลักฐานที่เชื่อมโยง
พยานคนแรก รายงานว่าเห็นชายที่ถือกระเป๋าฟิตเนสที่มีหมายเลขสมาชิกขึ้นต้นด้วย

1.   พยานคนแรก รายงานว่าเห็นชายที่ถือกระเป๋าฟิตเนสที่มีหมายเลขสมาชิกขึ้นต้นด้วย "48Z" และหนีไปขึ้นรถที่มีแผ่นป้ายทะเบียนที่มี "H42W" อยู่ในนั้น
2.   รายกพยานคนที่สอง (Annabel) ระบุว่าจำคนร้ายได้ว่าเคยเห็นที่ฟิตเนสในวันที่ 9 มกราคม 2018 ซึ่งการตรวจสอบข้อมูลการเช็คอินพบว่า Jeremy Bowers เช็คอินที่ฟิตเนสในวันนั้นารในลิสต์

จากข้อมูลทั้งหมดนี้ Jeremy Bowers เป็นผู้ต้องสงสัยหลักในคดีฆาตกรรมที่เกิดขึ้นใน SQL City เมื่อวันที่ 15 มกราคม 2018.
